bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/handler/EcovacsVacuumHandler.java

```diff
@@ -0,0 +1,819 @@
+import org.openhab.core.i18n.TranslationProvider;
+import org.openhab.core.io.net.http.HttpUtil;
+import org.openhab.core.library.types.DateTimeType;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.PercentType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.RawType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.library.unit.SIUnits;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.ThingStatusInfo;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.openhab.core.thing.binding.builder.ThingBuilder;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.openhab.core.types.StateOption;
+import org.openhab.core.types.UnDefType;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link EcovacsVacuumHandler} is responsible for handling data and commands from/to vacuum cleaners.
+ *
+ * @author Danny Baumann - Initial contribution
+ */
+@NonNullByDefault
+public class EcovacsVacuumHandler extends BaseThingHandler implements EcovacsDevice.EventListener {
+
+    private final Logger logger = LoggerFactory.getLogger(EcovacsVacuumHandler.class);
+
+    private final TranslationProvider i18Provider;
+    private final LocaleProvider localeProvider;
+    private final EcovacsDynamicStateDescriptionProvider stateDescriptionProvider;
+    private final Bundle bundle;
+
+    private final SchedulerTask initTask;
+    private final SchedulerTask reconnectTask;
+    private final SchedulerTask pollTask;
+    private @Nullable EcovacsDevice device;
+
+    private @Nullable Boolean lastWasCharging;
+    private @Nullable CleanMode lastCleanMode;
+    private @Nullable CleanMode lastActiveCleanMode;
+    private Optional<String> lastDownloadedCleanMapUrl = Optional.empty();
+    private long lastSuccessfulPollTimestamp;
+    private int lastDefaultCleaningPasses = 1;
+    private final String serialNumber;
+
+    public EcovacsVacuumHandler(Thing thing, TranslationProvider i18Provider, LocaleProvider localeProvider,
+            EcovacsDynamicStateDescriptionProvider stateDescriptionProvider) {
+        super(thing);
+        this.i18Provider = i18Provider;
+        this.localeProvider = localeProvider;
+        this.stateDescriptionProvider = stateDescriptionProvider;
+        bundle = FrameworkUtil.getBundle(getClass());
+        serialNumber = getConfigAs(EcovacsVacuumConfiguration.class).serialNumber;
```

Comment:
The correct place to initialize serialNumber is in `initialize()`. Otherwise it will not be picked up when changed. In the constructor you could even just set it to an empty string to satisfy null annotations.

Reply:
Exactly.